File: roman-to-integer/cs/Roman.cs
public class Roman {

    private static int NumeralToInt (string numeral) {
        switch (numeral.ToUpper()) {
            case "I": return 1;
            case "V": return 5;
            case "X": return 10;
            case "L": return 50;
            case "C": return 100;
            case "D": return 500;
            case "M": return 1000;
            default: return 0;
        }
    }

    public static int ToInt (string numerals) {
        int total = 0;
        int i = 0;

        while (i < numerals.Length) {
            string currentChar = numerals[i].ToString();
            string nextChar = "A";  // default A so it counts as 0

            if (i + 1 <= numerals.Length - 1) {
                nextChar = numerals[i + 1].ToString();
            }

            if (currentChar == "I" && nextChar == "V") { total = total + 4; i++; }
            else if (currentChar == "I" && nextChar == "X") { total = total + 9; i++; }
            else if (currentChar == "X" && nextChar == "L") { total = total + 40; i++; }
            else if (currentChar == "X" && nextChar == "C") { total = total + 90; i++; }
            else if (currentChar == "C" && nextChar == "D") { total = total + 400; i++; }
            else if (currentChar == "C" && nextChar == "M") { total = total + 900; i++; }
            else { total = total + NumeralToInt(currentChar); }
            
            i++;
        }

        return total;
    }
}

File: roman-to-integer/fs/Roman.fs
module Roman

let private NumeralToInt =
    function
    | "I" -> 1
    | "V" -> 5
    | "X" -> 10
    | "L" -> 50
    | "C" -> 100
    | "D" -> 500
    | "M" -> 1000
    | _ -> 0

let ToInt (numeral:string) : int =
    let mutable total = 0
    let mutable i = 0

    while i < numeral.Length do
        let currentChar = numeral.[i].ToString()
        // if nextChar does not exist use A because A will result in 0
        let nextChar : string =
            try
                if i + 1 <= numeral.Length then
                    numeral.[i + 1].ToString()
                else
                    "A"
            with _ ->
                "A"

        match currentChar, nextChar with
        // I can be placed before V (5) and X (10) to make 4 and 9. 
        | "I", "V" -> total <- total + 4; i <- i + 2;
        | "I", "X" -> total <- total + 9; i <- i + 2;
        // X can be placed before L (50) and C (100) to make 40 and 90. 
        | "X", "L" -> total <- total + 40; i <- i + 2;
        | "X", "C" -> total <- total + 90; i <- i + 2;
        // C can be placed before D (500) and M (1000) to make 400 and 900.
        | "C", "D" -> total <- total + 400; i <- i + 2;
        | "C", "M" -> total <- total + 900; i <- i + 2;
        | _ ->
            total <- total + NumeralToInt currentChar
            i <- i + 1

    total
